### ISSUE TITLE
How to run fairseq distributed mode in multiple nodes scenario?

### ISSUE THREAD

**Issue by asispatra** [pytorch/fairseq#463]
I am able to run fairseq translation example distributed mode in a single node. Can someone please tell me how run this across multiple node?

**Comment by myleott** [comment]
Here is the Distributed training section of the docs: https://fairseq.readthedocs.io/en/latest/getting_started.html#distributed-training

Closing for now, please reopen if you still have questions!

**Comment by ultrons** [comment]
Is example given at https://fairseq.readthedocs.io/en/latest/getting_started.html#distributed-training, expected to work for single node scenario?

I am running it on a machine with 8 V100 GPUs. I see it spawns 15 processes (rank 0 to rank 14), Shouldn't it be 8 processes only?

Following is the command line I am using:
python -m torch.distributed.launch --nproc_per_node=8 \
    --nnodes=1 --node_rank=0 --master_addr="10.138.0.6" \
    --master_port=8085 \
    $(which fairseq-train) /home/jupyter/data/wmt18_en_de_bpej32k \
    --arch transformer_vaswani_wmt_en_de_big --share-all-embeddings \
    --optimizer adam --adam-betas '(0.9, 0.98)' --clip-norm 0.0 \
    --lr-scheduler inverse_sqrt --warmup-init-lr 1e-07 --warmup-updates 4000 \
    --lr 0.0005 --min-lr 1e-09 \
    --dropout 0.3 --weight-decay 0.0 --criterion label_smoothed_cross_entropy --label-smoothing 0.1 \
    --max-tokens 3584 \
    --fp16

Are there some default assumptions/minimum number of nodes to run this?

**Comment by myleott** [comment]
I'm not sure why it launches 15 processes. But for a single node you can just run fairseq-train directly without torch.distributed.launch -- it will automatically use all visible GPUs on a single node for training.

**Comment by mortonjt** [comment]
I'm seeing something similar - when running on two nodes, I see 7 processes on each (rank (0-6) and rank (4-10)).  I'm running this on two separate nodes

```
ip=`curl ifconfig.me`
TOTAL_UPDATES=100000    # Total number of training steps
WARMUP_UPDATES=10000    # Warmup the learning rate over this many updates
PEAK_LR=0.0006          # Peak learning rate, adjust as needed
TOKENS_PER_SAMPLE=1024  # Max sequence length
MAX_POSITIONS=1024      # Num. positional embeddings (usually same as above)
MAX_SENTENCES=16        # Number of sequences per batch (batch size)
UPDATE_FREQ=8           # Increase the batch size 8x
DATA_DIR=<my data dir>
SAVE_DIR=<my checkpoint dir>
TB_DIR=<my logging dir>

python -m torch.distributed.launch --nproc_per_node=4 \
    --nnodes=2 --node_rank=1 --master_addr=$ip \
    --master_port=1234 \
    $(which fairseq-train) --fp16 $DATA_DIR \
    --task masked_lm --criterion masked_lm \
    --arch roberta_base --sample-break-mode complete --tokens-per-sample $TOKENS_PER_SAMPLE \
    --optimizer adam --adam-betas '(0.9, 0.98)' --adam-eps 1e-6 --clip-norm 0.0 \
    --lr-scheduler polynomial_decay --lr $PEAK_LR --warmup-updates $WARMUP_UPDATES --total-num-update $TOTAL_UPDATES \
    --dropout 0.1 --attention-dropout 0.1 --weight-decay 0.01 \
    --max-sentences $MAX_SENTENCES --update-freq $UPDATE_FREQ \
    --max-update $TOTAL_UPDATES --log-format tqdm --log-interval 1 \
    --save-interval-updates 1 \
    --keep-interval-updates 3 \
    --tensorboard-logdir $TB_DIR \
    --save-dir $SAVE_DIR
```

Furthermore, there aren't any logs / checkpoints -- have you seen something like this before?
